### ISSUE TITLE
Bug: `indent` is not reporting on wrong indentation

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#15509]
### Environment

Node version: latest
npm version: latest
Local ESLint version: v8.6.0
Global ESLint version: n/a
Operating System: Mac

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
indent: [2, "tab"]
```
</details>

See [demo](https://eslint.org/demo#eyJ0ZXh0IjoiJ3VzZSBzdHJpY3QnO1xuXG4vLyBodHRwczovLzI2Mi5lY21hLWludGVybmF0aW9uYWwub3JnLzExLjAvI3NlYy1pc3VuY2xhbXBlZGludGVnZXJlbGVtZW50dHlwZVxuXG4vKiBlc2xpbnQgaW5kZW50OiBbMiwgXCJ0YWJcIl0gKi9cblxubW9kdWxlLmV4cG9ydHMgPSBmdW5jdGlvbiBJc1VuY2xhbXBlZEludGVnZXJFbGVtZW50VHlwZSh0eXBlKSB7XG5cdHJldHVybiB0eXBlID09PSAnSW50OCdcbiAgICAgICAgfHwgdHlwZSA9PT0gJ1VpbnQ4J1xuICAgICAgICB8fCB0eXBlID09PSAnSW50MTYnXG4gICAgICAgIHx8IHR5cGUgPT09ICdVaW50MTYnXG4gICAgICAgIHx8IHR5cGUgPT09ICdJbnQzMidcbiAgICAgICAgfHwgdHlwZSA9PT0gJ1VpbnQzMic7XG59O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1ub25vY3RhbC1kZWNpbWFsLWVzY2FwZSI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW5zYWZlLW9wdGlvbmFsLWNoYWluaW5nIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjIsImluZGVudCI6Mn0sImVudiI6eyJub2RlIjp0cnVlfX19)

```js
'use strict';

// https://262.ecma-international.org/11.0/#sec-isunclampedintegerelementtype

/* eslint indent: [2, "tab"] */

module.exports = function IsUnclampedIntegerElementType(type) {
	return type === 'Int8'
        || type === 'Uint8'
        || type === 'Int16'
        || type === 'Uint16'
        || type === 'Int32'
        || type === 'Uint32';
};
```
Note how the `return` correctly has a tab, but the `||` lines all have spaces. eslint does not warn on this.

### What did you expect to happen?

eslint should warn on spaces being used for indentation, when configured to "tab".

### What actually happened?

No warning.

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

This definitely happened in eslint 7 also, and probably older versions.

**Comment by nzakas** [comment]
It looks like this behavior is intentional:
https://github.com/eslint/eslint/blob/670f48113708b556e8ba870607766cb4707791cf/lib/rules/indent.js#L1126-L1130

This seems like an acknowledgment that this is a bug, but one that remains just for backwards compatibility with the original rule. So while we can’t change the default, I think it’s reasonable to add an option that opts-in to a fixed behavior.

@eslint/eslint-tsc thoughts?

**Comment by ljharb** [comment]
Awesome, an option sounds great too!

As soon as I get a pointer and the go-ahead I'd be happy to take a crack at a PR.

**Comment by mdjermanovic** [comment]
This was discussed several times as an enhancement to this rule (e.g., https://github.com/eslint/eslint/issues/12248, https://github.com/eslint/eslint/issues/12255, https://github.com/eslint/eslint/issues/12427), but it didn't reach a consensus.

Since this wasn't considered a bug, I think it falls under our policy of not adding new options to stylistic rules.

**Comment by ljharb** [comment]
@mdjermanovic to me it’s clearly a bug - the entire point of the indent rule is to ensure proper indentation of *everything*, no exceptions, and it allowed me to have some lines indented with spaces in a tabbed codebase. I don’t care if the fix is by default or under an option for compat, but i don’t think it’s reasonable to say this falls under that policy, since the current behavior defeats the entire purpose of the rule since its inception.

**Comment by btmills** [comment]
The first version of indent was just about blocks and statements. We’ve since added on options for a whole bunch of multi line expressions. It seems reasonable to me that at some point along the way, the rule’s scope expanded to include all lines, not just blocks, statements, and a couple dozen other cases, which would make this an omission and plausibly justifiable a bug. Given that, I’m with @nzakas.

**Comment by nzakas** [comment]
Ok, marking as accepted. @ljharb feel free to open a PR.

**Comment by ljharb** [comment]
Got #15555 up; please review :-)

**Comment by snitin315** [comment]
I'll pick this.

**Comment by nzakas** [comment]
Hi folks, I just wanted to let everyone know that the ESLint team has decided to deprecate all formatting rules, including indent, and recommend people use source code formatters like [Prettier](https://prettier.io) or [dprint](https://dprint.dev) instead. As a result, we are closing all outstanding issues related to these rules. For more information, please see https://github.com/eslint/eslint/issues/17522. Thanks.

**Comment by ljharb** [comment]
That’s very sad to hear.
